pkg/integrations/v2/app_o11y_receiver/sourcemaps/sourcemaps.go

```diff
@@ -0,0 +1,331 @@
+	Get(url string) (resp *http.Response, err error)
+}
+
+// FileService is interface for a service that can be used to load source maps
+// from file system
+type FileService interface {
+	Stat(name string) (fs.FileInfo, error)
+	ReadFile(name string) ([]byte, error)
+}
+
+type osFileService struct{}
+
+func (s *osFileService) Stat(name string) (fs.FileInfo, error) {
+	return os.Stat(name)
+}
+
+func (s *osFileService) ReadFile(name string) ([]byte, error) {
+	return os.ReadFile(name)
+}
+
+var reSourceMap = "//[#@]\\s(source(?:Mapping)?URL)=\\s*(?P<url>\\S+)\r?\n?$"
+
+type sourceMap struct {
+	consumer *sourcemap.Consumer
+}
+
+type sourceMapMetrics struct {
+	cacheSize *prometheus.CounterVec
+	downloads *prometheus.CounterVec
+	fileReads *prometheus.CounterVec
+}
+
+// RealSourceMapStore is an implementation of SourceMapStore
+// that can download source maps or read them from file system
+type RealSourceMapStore struct {
+	sync.Mutex
+	l           log.Logger
+	httpClient  HTTPClient
+	fileService FileService
+	config      config.SourceMapConfig
+	cache       map[string]*sourceMap
+	metrics     *sourceMapMetrics
+}
+
+// NewSourceMapStore creates an instance of SourceMapStore.
+// httpClient and fileService will be instantiated to defaults if nil is provided
+func NewSourceMapStore(l log.Logger, config config.SourceMapConfig, reg *prometheus.Registry, httpClient HTTPClient, fileService FileService) SourceMapStore {
+	if httpClient == nil {
+		httpClient = &http.Client{
+			Timeout: config.DownloadTimeout,
+		}
+	}
+
+	if fileService == nil {
+		fileService = &osFileService{}
+	}
+
+	metrics := &sourceMapMetrics{
+		cacheSize: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "app_o11y_receiver_sourcemap_cache_size",
+			Help: "number of items in sourcemap cache, per origin",
+		}, []string{"origin"}),
+		downloads: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "app_o11y_receiver_sourcemap_downloads",
+			Help: "downloads by the sourcemap service",
+		}, []string{"origin", "http_status"}),
+		fileReads: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "app_o11y_receiver_sourcemap_file_reads",
+			Help: "",
+		}, []string{"origin", "status"}),
+	}
+	reg.MustRegister(metrics.cacheSize, metrics.downloads, metrics.fileReads)
+
+	return &RealSourceMapStore{
+		l:           l,
+		httpClient:  httpClient,
+		fileService: fileService,
+		config:      config,
+		cache:       make(map[string]*sourceMap),
+		metrics:     metrics,
+	}
+}
+
+func (store *RealSourceMapStore) downloadFileContents(url string) ([]byte, error) {
+	resp, err := store.httpClient.Get(url)
+	if err != nil {
+		store.metrics.downloads.WithLabelValues(getOrigin(url), "?").Inc()
+		return nil, err
+	}
+	store.metrics.downloads.WithLabelValues(getOrigin(url), fmt.Sprint(resp.StatusCode)).Inc()
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("unexpected status %v", resp.StatusCode)
+	}
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
```

Comment:
Maybe we can check content-type as well here to make sure we do not download non javascript files

Reply:
I'm not sure that it's safe to limit only to `.js`. It could be at least `wasm` too?

Reply:
I guess we have a bit too many types indeed. I am bit extra careful about sanitising/validating everything as we load resources from the filesystem/internet based on user data